cpp/src/arrow/compute/exec.h

```diff
@@ -240,6 +246,8 @@ struct ARROW_EXPORT ExecBatch {
 
   ExecBatch Slice(int64_t offset, int64_t length) const;
 
+  Result<ExecBatch> SelectValues(const std::vector<int>& ids) const;
```

Comment:
What is this for?

Reply:
It is [used](https://github.com/apache/arrow/pull/34311/files#diff-f53040e5fc284581263737f96238c8a76ef54d13f01e34305c1d1d9de5e6e34bR214) for selecting the values (i.e., columns) of the batch that correspond to the segment-keys. The resulting sub-batch, having just these values, is [processed by the segmenter](https://github.com/apache/arrow/pull/34311/files#diff-f53040e5fc284581263737f96238c8a76ef54d13f01e34305c1d1d9de5e6e34bR216).

Reply:
I'd imagine "extract" keys is needed for aggregation without the segmenter change, how was that done before?

Reply:
Regular aggregation works differently - it uses encoders that compute a hash for the selected keys. The segmenter doesn't use encoders.

Reply:
"Getting key fields out of a batch" seems something that is universally needed for any operator that need "key" that's why I found it surprising that we needed to add such functionality for segment aggregation. Seems like we added yet another way to "extract keys" (we needed to extract keys also in, for example, asof join, which is done also differently).

Probably can be follow up but I do feel there are too many ways to extract key from exec batches.

Reply:
If this is an issue, I agree it should be a follow-up one.

Reply:
Please add a follow-up issue